stubs/mu-plugin.php

```diff
@@ -0,0 +1,357 @@
+            $bytes,
+            function_exists('\get_num_queries') ? \get_num_queries() : null,
+            $requestStart ? round((microtime(true) - $requestStart) * 1000, 2) : null
+        );
+
+        if ($redisSample) {
+            $metrics .= ' ' . implode(' ', array_map(static function ($metric, $value) {
+                return "sample#{$metric}={$value}";
+            }, array_keys($redisSample), $redisSample));
+        }
+
+        if ($relaySample) {
+            $metrics .= ' ' . implode(' ', array_map(static function ($metric, $value) {
+                return "sample#{$metric}={$value}";
+            }, array_keys($relaySample), $relaySample));
+        }
+
+        return $metrics;
+    }
+
+    protected static function calculateHitRatio(int $hits, int $misses): float
+    {
+        $total = $hits + $misses;
+
+        return $total > 0 ? round($hits / ($total / 100), 1) : 100;
+    }
+
+    protected static function calculateBytes(array $cache)
+    {
+        $bytes = array_map(function ($keys) {
+            return strlen(serialize($keys));
+        }, $cache);
+
+        return array_sum($bytes);
+    }
+
+    protected static function mapRedisInfo(array $info, int $database): array
+    {
+        $total = intval($info['keyspace_hits'] + $info['keyspace_misses']);
+
+        $dbKey = "db{$database}";
+
+        if (isset($info[$dbKey])) {
+            $keyspace = array_column(array_map(static function ($value) {
+                return explode('=', $value);
+            }, explode(',', $info[$dbKey])), 1, 0);
+
+            $keys = intval($keyspace['keys']);
+        }
+
+        return [
+            'redis-hits' => $info['keyspace_hits'],
+            'redis-misses' => $info['keyspace_misses'],
+            'redis-hit-ratio' => $total > 0 ? round($info['keyspace_hits'] / ($total / 100), 2) : 100,
+            'redis-ops-per-sec' => $info['instantaneous_ops_per_sec'],
+            'redis-evicted-keys' => $info['evicted_keys'],
+            'redis-used-memory' => $info['used_memory'],
+            'redis-used-memory-rss' => $info['used_memory_rss'],
+            'redis-memory-fragmentation-ratio' => $info['mem_fragmentation_ratio'] ?? 0,
+            'redis-connected-clients' => $info['connected_clients'],
+            'redis-tracking-clients' => $info['tracking_clients'] ?? 0,
+            'redis-rejected-connections' => $info['rejected_connections'],
+            'redis-keys' => $keys ?? 00,
+        ];
+    }
+
+    protected static function mapRelayStats(array $stats, int $database): array
+    {
+        $total = intval($stats['stats']['hits'] + $stats['stats']['misses']);
+
+        return [
+            'relay-hits' => $stats['stats']['hits'],
+            'relay-misses' => $stats['stats']['misses'],
+            'relay-hit-ratio' => $total > 0 ? round($stats['stats']['hits'] / ($total / 100), 2) : 100,
+            'relay-ops-per-sec' => $stats['stats']['ops_per_sec'],
+            'relay-keys' => '',
```

Comment:
Yeah, that's tricky. Check out `RelayMetrics.php` in Object Cache Pro.

Reply:
I have, but the approach in that class doesn't seem to work. In my testing, `$keys` is always null because the connection doesn't have a `keys` property.

https://github.com/rhubarbgroup/object-cache-pro/blob/develop/src/Metrics/RelayMetrics.php#L94